[6주차] 레스토랑 상세 페이지 구현하기

hanurii opened the pull request:
- 저와의 약속으로 PR만 먼저 보내놓겠습니다.

Commit:
project init

Commit:
모든 경로가 index.html로 이동하도록 webpack 설정 변경

- historyApiFallback 설정을 추가하여 모든 경로가 한 곳으로 모이도록 함

Commit:
RestaurantsPage에 대한 테스트를 작성한다

Commit:
RestaurantsPage를 구현한다

Commit:
RestaurantsPage를 리팩토링한다

- App, App.test 의 모든 로직을 RestaurantsPage로 옮겼기 때문에 불필요한 로직 제거

Commit:
HomePage에 대한 테스트를 작성한다

Commit:
HomePage를 구현한다

Commit:
HomePage를 리팩토링한다

- App 컴포넌트에서 HomePage 컴포넌트 분리

Commit:
NotFoundPage에 대한 테스트를 작성한다

Commit:
NotFoundPage를 구현한다

Commit:
NotFoundPage를 리팩토링한다

- App 컴포넌트에서 NotFoundPage 컴포넌트 분리

Commit:
AboutPage에 대한 테스트를 작성한다

Commit:
AboutPage를 구현한다

Commit:
AboutPage를 리팩토링한다

Commit:
About, Restaurants가 화면에 보이는 테스트를 작성한다

Commit:
About, Restaurants가 화면에 보이도록 구현한다

Commit:
HeaderPage에 대한 테스트를 작성한다

Commit:
HeaderPage를 구현한다

Commit:
Restaurants, About, Home, NotFound에 헤더 추가에 대한 테스트를 작성한다

Commit:
Restaurants, About, Home, NotFound에 헤더 추가를 구현한다

Commit:
About 페이지 문구에 대한 테스트를 작성한다

Commit:
About 페이지 문구를 구현한다

Commit:
About 페이지 제목에 대한 테스트를 작성한다

Commit:
About 페이지 제목을 구현한다

Commit:
Home 페이지 제목에 대한 테스트를 작성한다

Commit:
Home 페이지 제목을 구현한다

Commit:
헤더를 누르면 홈으로 이동하는 동작에 대한 테스트를 작성한다

Commit:
헤더를 누르면 홈으로 이동하는 동작을 구현한다

Commit:
홈페이지에서 About, Restaurants 페이지로 이동하는 a 태그에 대한 테스트를 작성한다

- 테스트가 미흡하여 구현된 부분 테스트 작성 추가

Hamill210 commented:
### 과제 진행률
- TDD로 기능 구현
- 커버리지 100% 달성 (94% 달성)
- About 페이지 구현 (완)
- Not Found 페이지 구현 (완)
- 레스토랑 상세페이지 구현 (미완)

### To-do
- App 컴포넌트에서 NotFound로 가는 경우에 대한 테스트 추가
- React Router 도입
- 레스토랑 상세페이지 구현
- 지난 주 미흡했던 비동기 구현, 테스트 재점검

Commit:
React Router 설치

Commit:
App 컴포넌트 Router 적용에 대한 테스트를 작성한다

Commit:
App 컴포넌트 Router 적용을 구현한다

Commit:
App 컴포넌트에서 라우팅 처리에 대한 테스트를 작성한다

Commit:
App 컴포넌트에서 라우팅 처리를 구현한다

Commit:
App 컴포넌트에서 라우팅 처리를 리팩토링한다

Commit:
공통 헤더에 대한 테스트를 작성한다

Commit:
공통 헤더를 구현한다

Commit:
공통 헤더를 리팩토링한다

Commit:
AboutPage 테스트 잘못된 이름 정정

Commit:
공통 헤더를 누르면 홈페이지로 이동에 대한 테스트를 작성한다

Commit:
공통 헤더를 누르면 홈페이지로 이동을 구현한다

Commit:
홈페이지에서 다른 페이지로 이동하는 링크에 대해 테스트를 작성한다

Commit:
홈페이지에서 다른 페이지로 이동하는 링크를 구현한다

Commit:
eslint 적용

Hamill210 commented:
### 과제 진행률
- TDD로 기능 구현
- ~커버리지 100% 달성 (완)~
- ~About 페이지 구현 (완)~
- ~Not Found 페이지 구현 (완)~
- 레스토랑 상세페이지 구현 (미완)

### To-do
- 레스토랑 상세페이지 구현
- 지난 주 미흡했던 비동기 구현, 테스트 재점검
- 지난 주 반영하지 못했던 피드백 반영

moonkii commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,15 @@
+    <AboutPage />
+  ));
+
+  expect(getByText('About')).not.toBeNull();
+  expect(getByText('About')).toContainHTML('h2');
+
+  expect(getByText('About 페이지 입니다.')).not.toBeNull();
```
`getBy~` 와   `queryBy~` 의 차이점을 알고 계신가요?

각각 언제 사용하면 좋을까요?

hanurii replied:
잘 모르던 차이점을 공부했습니다.
- getBy : 쿼리에 대해 일치하는 노드를 반환하고 일치하는 요소가 없거나 둘 이상의 일치가 발견되면 descriptive 에러를 발생시킵니다.
- queryBy : 쿼리에 대해 일치하는 노드를 반환하고 일치하는 요소가 없으면 null을 반환합니다. 존재하지 않는 엘리먼트를 선언하는 데 유용합니다.

반드시 일치하는 노드가 존재해야 할 경우 getBy로 나타내주고, 존재해도되고 하지 않아도 되는 노드의 경우 queryBy를 사용하면 좋을 것 같습니다.

moonkii replied:
그렇기 떄문에 `toBeNull()` 을 사용하는 경우 결과를 null 을 반환하는 `queryBy~` 를 사용하는게 더 좋겠죠?

hanurii replied:
아 그렇네요! 근데 저는 차이점을 공부하면서 든 생각이 
1. getBy 해당 문구가 반드시 있어야 함
2. queryBy 해당 문구가 있어도되고 없어도되고 

이 부분에 집중했었는데, 기환님은 matcher 를 보며 생각하신 듯합니다.
또 다른 시야를 보게 해주셔서 감사합니다 ㅎㅎ 한번 반영해보겠습니다

moonkii commented on src/App.jsx:
```diff
@@ -1,32 +1,27 @@
-import CategoriesContainer from './CategoriesContainer';
-import RestaurantsContainer from './RestaurantsContainer';
-
+import React from 'react';
```
React 17 버전부터는 `import React from 'react';` 를 생략해줄 수 있어요.

업데이트 내용 등도 최신 버전에 맞게 따라가는게 중요해요.
리액트 공식 문서를 참조하시면  확인하실 수 있어요!

참고:
- https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

hanurii replied:
> With the new transform, you can use JSX without importing React.

오.. 생략해줄수 있는 사실을 처음 알았습니다. 업데이트 되는 공식 문서도 잘 챙겨보겠습니다.

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,68 @@
-  const dispatch = jest.fn();
+describe('App', () => {
+  beforeEach(() => {
+    const dispatch = jest.fn();
```
`dispatch` 를 외부에서 선언해야 테스트할 때 호출이 가능할거에요.

hanurii replied:
아 생각해보니 그렇네요 ㅎㅎ.. 수정해서 반영하겠습니다. 피드백 감사합니다!

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,68 @@
+    expect(getByText('헤더').closest('a')).toHaveAttribute('href', '/');
+  });
+
+  it('renders Home', () => {
+    const { container } = renderApp({ path: '/' });
+
+    expect(container).toHaveTextContent('Home');
+  });
+
+  it('renders About', () => {
+    const { container } = renderApp({ path: '/about' });
 
```
테스트가 path 따라 시나리오가 달라질때 사용하는게 `context` 에요.
찾아보시면 도움되실거에요 ;)

hanurii replied:
아 path 따라 context를 달리 해주는군요...! 한 번 찾아보고 반영해보겠습니다. 피드백 감사합니다!

moonkii commented on src/NotFoundPage.test.jsx:
```diff
@@ -0,0 +1,10 @@
+
+import NotFoundPage from './NotFoundPage';
+
+test('NotFoundPage', () => {
+  render((
+    <NotFoundPage />
+  ));
+});
```
웬만하면 `test` 보다는 `describe - (context) - it` 형태로 작성해주시는게 좋아요.

`NotFoundPage` 에서는 어떤걸 테스트해주면 좋을까요?

hanurii replied:
넵 describe - it 형태로 변경해서 반영하겠습니다. 
NotFound 에선 '404 Not Found' 문자열이 존재하는지에 대한 테스트를 해주면 좋을 것 같아요!

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/121040363-ccf8d180-c7ec-11eb-91ce-d7fc1050a3fa.png)

약속을 지키기 위해 PR을 보내시는 모습 멋진것 같아요 👍

Commit:
import 모듈에서 React 제거

- React 17부터는 React 모듈을 import 해주지 않아도 JSX를 사용 할 수 있다고 함

Commit:
dispatch를 사용 할 수 있도록 beforeEach에서 꺼낸다

- beforeEach 안에 dispatch가 선언되면  다른 곳에서 사용 할 수 없음

Commit:
path 따라 시나리오가 달라지는 경우 context로 감싸준다

Commit:
NotFound 컴포넌트 테스트 추가

- 원하는 문구가 존재하는지 확인
- 해당 HTML 엘리먼트를 사용하는지 확인

Commit:
레스토랑 목록을 클릭하면 다른 링크로 이동에 대한 테스트를 작성한다

Commit:
레스토랑 목록을 클릭하면 다른 링크로 이동을 구현한다

Commit:
레스토랑 목록을 클릭하면 다른 링크로 이동을 리팩토링한다

Commit:
App 컴포넌트에 레스토랑 상세정보 Route 추가에 대한 테스트를 작성한다

Commit:
App 컴포넌트에 레스토랑 상세정보 Route 추가를 구현한다

Commit:
App 컴포넌트에 레스토랑 상세정보 Route 추가를 리팩토링한다

Commit:
RestaurantDetailPage에서 레스토랑 정보를 가져오는 것에 대한 테스트를 작성한다

Commit:
RestaurantDetailPage에서 레스토랑 정보를 가져오는 것을 구현한다

Commit:
리액트 라우터 새로고침 시 404 나타나는 버그 수정

Commit:
Eslint 적용

Hamill210 commented:
### 과제 진행률
- TDD로 기능 구현
- ~커버리지 100% 달성 (완)~
- ~About 페이지 구현 (완)~
- ~Not Found 페이지 구현 (완)~
- ~레스토랑 상세페이지 구현 (완)~
- ~인수 테스트 통과~

### To-do
- 지난 주 미흡했던 비동기 구현, 테스트 재점검
- 지난 주 반영하지 못했던 피드백 반영
- 다양한 라우터 테스트 시도해보기

Commit:
식당 상세정보에서 메뉴가 나타나지 않는 문제 해결

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,92 @@
+    }
+
+    context('with Home page', () => {
+      it('verifies header', () => {
```
`verifies` 라는 표현은 무슨 의미인가요?

hanurii replied:
검증하다 라는 의미로 사용한 단어입니다. 적절치 못한 단어 사용이었을까요? ㅜ

hanurii replied:
renders라는 표현이 더 적절한 것 같습니다!

moonkii replied:
describe 에 작성한 대상을 `it` 에 대입하셔서 테스트 시나리오가 성립하는지 보시면 좋아요.

`App verifies header` 라고 한다면 App 이 헤더를 검증한다는건데 시나리오 자체가 성립하지 않는 말이겠죠?
App 이 헤더를 render 한다는걸 확인하고 싶은게 테스트 내용일 거에요.

hanurii replied:
넵 대입해서 시나리오 성립하는지 확인하는 것 잊지 않겠습니다!

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,92 @@
-  const { queryByText } = render((
-    <App />
-  ));
+    context('with About page', () => {
```
페이지는 나타나는 결과에요.
테스트 맥락상 어떤게 구분되어지고 있나요?

hanurii replied:
https://github.com/CodeSoom/react-week6-assignment-1/pull/50#discussion_r646668084
이 코멘트를 보고 context로 구분지어 줬는데, 제가 잘못 받아들인 것 같아요. 구분짓는 게 아니라 describe로 설명되어져야 할 것 같은데 제가 생각한 부분이 맞나요?

hanurii replied:
```javascript
describe('AboutPage', () => {
  it('renders', () => {})
})
```

형태로 가는 게 좋아보입니다...! 어떻게 생각하시나요

moonkii replied:
`context` 를 사용하는건 적절해 보여요.
다만, 어떤 맥락상의 차이가 있는지 드러내는게 중요하겠죠?

만약 저라면 path 에 따라 테스트 시나리오가 달라지기 때문에
`context('when path is /about ...` 같이 어떤 맥락인지 구체적으로 명시해줬을 것 같아요.

hanurii replied:
오... 적절한 예시를 들어주셔서 정말 감사합니다!! 한 번 반영해보겠습니다

moonkii commented on src/NotFoundPage.test.jsx:
```diff
@@ -0,0 +1,14 @@
+    ));
+
+    expect(getByText('404 Not Found')).not.toBeNull();
+    expect(getByText('404 Not Found')).toContainHTML('p');
```
태그까지 테스트를 해주는 이유가있나요?
어떤걸 테스트하고 싶고, 어떤 의도를 드러내고 싶은가요?

hanurii replied:
해당 문구의 태그를 테스트 함으로써 테스트만 봐도 전체적인 그림이 그려지도록 하고싶어서 태그를 테스트했습니다.
혹시 배드 케이스일까요?

moonkii replied:
의미있는 HTML 마크업을 사용하는게 중요하긴하지만 테스트에서 확인하기보단 이 부분은 직접 코드를 통해 확인하는게 좋을 것 같아요.
그렇지 않으면 테스트코드에 직접 HTML 마크업을 다 드러내줘야하는 상황이 발생하겠죠?

hanurii replied:
저도 일부만 드러내야할지 모두 드러내야할지 많은 고민을 했었는데, 명쾌히 답을 내려주셔서 감사합니다!
지금 저는 마크업과 그 속성까지 테스트를 하고 있는데, 혹시 이런 테스트는 어떤 경우에 해주는 게 좀더 적절할까요?

moonkii replied:
예를들어 이미지에 관한 거나 링크에 관한 것에 대해 특별하게 표현해주고 싶을 때 해줄 수 있겠죠?

moonkii commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,36 @@
+      </p>
+      <h3>메뉴</h3>
+      <ul>
+        {menuItems.map(({ id, name: menuName }) => (
+          <li key={id}>
+            {menuName}
+          </li>
+        ))}
```
메뉴 아이템이 없을 때는 어떻게 보여주면 좋을까요? 🤔

hanurii replied:
음.. 뭔가 없을 경우를 대비한 텍스트가 있으면 좋을 것 같아요. 한 번 생각해보고 반영해보겠습니다.

moonkii commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,43 @@
+
+describe('RestaurantDetailPage', () => {
+  it('renders', () => {
+    const dispatch = jest.fn();
+
+    useDispatch.mockImplementation(() => dispatch);
+
+    useSelector.mockImplementation((selector) => selector({
+      regions: [],
+      categories: [],
+      restaurants: [],
+      restaurant: {
+        id: 1,
+        categoryId: 1,
+        name: '양천주가',
+        address: '서울 강남구',
+        menuItems: [
+          { id: 1, restaurantId: 1, name: '비빔밥' },
+        ],
+      },
+    }));
```
이부분은 웬만하면 `beforeEach` 에서 처리해주시는게 좋아요.
매 테스트 마다 공통적로 필요한 부분이니까요.
모든 코드에는 의도가 드러나도록 작성하시면 좋아요.

hanurii replied:
넵 감사합니다. 혹시 beforeEach가 아니고 beforeAll에다가 넣어주면 테스트 중간에 값이 변경되면 그 값으로 계속 공유되서 테스트가 진행될까요?

moonkii replied:
`beforeEach` 와 `beforeAll` 의 차이가 무엇인가요?

hanurii replied:
- beforeEach : 매 테스트 시작 전 실행
- beforeAll : 전체 테스트 시작 전 1번 실행

쓰임새는 알고 있지만 선언된 값이 공유되는지 알지 못해서 여쭤봤습니다. 이 부분 좀더 공부해보겠습니다.

moonkii commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,43 @@
+      },
+    }));
+
+    const match = {
+      params: {
+        restaurantId: 1,
+      },
+    };
```
이 부분도 테스트가 여러개라면 공통적으로 사용될 수 있을 것 같은데 어디서 선언해주면 좋을까요?

hanurii replied:
~기환님 말씀대로면 beforeEach입니다 ㅎㅎ~

beforeEach에 넣으면 다른 곳에서 사용을 못하니... beforeEach 밖에 선언하는 게 맞을 것 같네요

moonkii commented on src/actions.test.js:
```diff
@@ -83,4 +85,20 @@ describe('actions', () => {
+      });
+    });
+
+    it('runs setRestaurant', async () => {
```
`loadRestaurant runs setRestaurant` 라고 하면 어떻게 이해가 되나요?
테스트 시나리오는 어떤걸 확인하고싶은지 고민해보시면 좋아요.

hanurii replied:
더 적절한 시나리오로 설명 되도록 수정해보겠습니다. 피드백 감사합니다!!

hanurii replied:
run이 아니라 set, insert 라는 표현이 더 적절할 것 같습니다

moonkii replied:
시나리오는 말그대로 프로그래머가 이해하기 위한 문장이에요.

그렇기 때문에 의미가 드러나는게 중요하지 꼭 코드명을 안써주셔도 괜찮아요.

말씀하신대로 `it sets restaurant` 같은 형태로 말이 읽히는게 중요해요.

hanurii replied:
읽히도록 만들기...! 명심하겠습니다!

moonkii commented on src/services/__mocks__/api.js:
```diff
@@ -9,3 +9,7 @@ export async function fetchCategories() {
 export async function fetchRestaurants() {
   return [];
 }
+
+export async function fetchRestaurant() {
+  return [];
+}
```
mock 파일은 mock function 으로 선언해주셔야 해요!

hanurii replied:
mock function ?? 그게 뭔가요?

hanurii replied:
여길 참고해보겠습니다

https://jestjs.io/docs/mock-functions

moonkii replied:
공부하실 때 공식문서를 꼭 읽어보세요!

moonkii commented:
코드숨 메인 브랜치에 PR을 잘못 보내셨어요!

Hamill210 commented:
> 코드숨 메인 브랜치에 PR을 잘못 보내셨어요!

수정했습니다!

Commit:
null 검증 시엔 queryBy를 사용

- toBeNull()이란 Matcher는 null이다 라는 걸 표현해주는 매처
- queryBy는 일치하는 문구가 없을 경우 Null을 리턴
- getBy보다 매처랑 어울리는 queryBy를 사용

Commit:
식당 상세정보에서 메뉴가 없을 경우에 대한 테스트를 작성한다

Commit:
식당 상세정보에서 메뉴가 없을 경우를 구현한다

Commit:
식당 상세정보에서 메뉴가 없을 경우를 리팩토링한다

Commit:
시나리오 설명이 자연스럽도록 테스트 수정

Commit:
시나리오 설명이 자연스럽도록 테스트 수정 2

Commit:
mock 파일에선 mock functions 을 사용

moonkii commented on index.html:
```diff
@@ -6,6 +6,6 @@
   <body>
     <div id="app"></div>
-    <script src="main.js"></script>
+    <script src="/main.js"></script>
```
`/` 있고 없고에 따라 어떤 차이가 발생할까요? (틀렸다는 말이 아니에요!)

hanurii replied:
절대경로와 상대경로의 차이라고 어렴풋이 추측만하고 정확히는 모르는 내용이었는데,
동우님이 슬랙에 올려주신 내용을 확인하니 추측한 부분이 맞았습니다.

절대경로는 어느 곳에 위치해 있어도 해당 경로는 고정값이기에 항상 같은 경로값을 내뱉고,
상대경로는 현재 있는 위치, 경로를 기준으로 값이 상대적으로 바뀌기에 위치에 따라 다른 경로값을 내뱉습니다.

moonkii commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,14 @@
+
+import AboutPage from './AboutPage';
+
+test('AboutPage', () => {
+  const { getByText, queryByText } = render((
+    <AboutPage />
+  ));
+
+  expect(queryByText('About')).not.toBeNull();
+  expect(getByText('About')).toContainHTML('h2');
+
+  expect(getByText('About 페이지 입니다.')).not.toBeNull();
+});
```
테스트 코드를 작성하실 때 웬만하면 describe ~ (context) ~ it 형태를 사용해주세요.
그러면 '왜' 테스트를 하는지, 어떤걸 테스트하고 싶은지 코드를 읽는 입장에서 더 알기가 쉬워져요.

hanurii replied:
넵! 지난 코멘트보고 일부만 수정해서 이 부분을 놓쳤네요.. 다시 반영하겠습니다~

moonkii commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,46 @@
+        {address}
+      </p>
+      <h3>메뉴</h3>
+      {getMenuItems()}
```
이렇게 함수로 분리해주시면 코드 가독성이 매우 떨어지는 것 같아요.
함수는 함수를 사용하는 곳에서의 인터페이스가 가장 중요해요.
이부분에 `getMenuItems` 라고 했을 때 어떤 의미가 드러나나요?

따로 함수로 분리하지 마시고 차라리 삼항연산자를 사용해보는건 어떨까요? :)

hanurii replied:
아항..! 이렇게하면 가독성이 떨어지는지 몰랐습니다... (회사 코드에도 이렇게 반영해버렸는데...!)
사실 이렇게 한 이유는 삼항 연산자로 하니 가독성이 떨어져보일 것 같아서 함수로 빼내어 주었는데요
아예 엘리먼트들을 감춰버린 상황 자체가 가독성이 떨어져보일 수 있는지 몰랐습니다.
다시 수정해보겠습니다. 피드백 감사합니다!!

moonkii commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,85 @@
+    ));
+  }
+
+  describe('dispatch', () => {
```
특별한 경우가 아니라면 `describe` 내부에 `describe` 를 사용하면 테스트 코드를 읽는 데도 복잡하고, 뭔가 잘 못 했을 확률이 높아요.

hanurii replied:
음.. 이 코드만 보면 굳이 describe를 쓰지 않고 it으로 표현해줬어도 됐을 것 같습니다. 
describe 내부에 describe를 사용하면 테스트가 복잡해지고, 잘못 했을 확률이 높아진다고 하셨는데요
혹시 왜 그런지... 좀 더 설명해주실 수 있나요?

moonkii replied:
describe 는 말그대로 테스트의 서술 대상이에요.
그래서 it 에다가 대입시켜서 테스트 시나리오를 읽어요.
그런데 특정 컴포넌트를 테스트하기 위해 describe 로 대상을 지정했는데 또 내부에 테스트를 따로 잡아주고 싶은
대상이 있다는건 뭔가 구조가 잘못되었을 확률이 높을 수 있겠죠? 
(무조건 틀렸다는 건 아니에요. 상황에 따라 다를수 있어요. 그래서 확률이라고 표현하는 거에요.)

가령 컴포넌트가 복잡해지다 보면 `context` 내부에 계속해서 `context` 가 들어가는 경우가 있는데
`if` 문 안에 계속 해서 `if` 가 중첩되서 들어가는 것과 마찬가지로 점점 조건이 읽기가 어려워질거에요.
그럴땐 구조적으로 변경해야할 신호일 확률이 높아요.

hanurii replied:
아하...! 무슨 말인지 이해했습니다.. 자세한 설명 정말 감사합니다!

moonkii commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,85 @@
+      }));
+    });
+
+    it('changes restaurant', () => {
```
`dispatch changes restaurant` 라고 하면 어떻게 의미가 해석되나요?

hanurii replied:
ㅎㅎ.. 제가 이야기하고자 하는 시나리오가 아니네요. 이런 부분 유의하면서 테스트를 작성해야 겠습니다. 피드백 감사합니다!!

moonkii commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,85 @@
+  });
+
+  context('with menu', () => {
+    beforeEach(() => {
+      useSelector.mockImplementation((selector) => selector({
+        restaurant: {
+          id: 1,
+          categoryId: 1,
+          name: '양천주가',
+          address: '서울 강남구',
+          menuItems: [
+            { id: 1, restaurantId: 1, name: '비빔밥' },
+          ],
+        },
+      }));
+    });
```
`menuItems` 만 다르고 반복되는데 어떻게 해줄 수 있을까요?
`given2` 라이브러리를 찾아보시면 도움이 되실거에요 :)

hanurii replied:
한 곳에서 관리 할 수 있도록 만들어보겠습니다...!

Commit:
path 경로에 따라 context 시나리오 작성

-  'when path is ${pathname}'  과 같은 형태의  시나리오로 작성
- describe가 아니라 context를 사용해서 여러 개의 경로를 나타 낼 수 있도록 만듦

Commit:
menuItems만 다르고 중복되는 코드 리팩토링

- 함수를 만들어 중복을 제거

Commit:
given2 라이브러리 사용하여 중복 코드 제거

Commit:
describe - (context) - it 구조로 테스트를 작성하라

- 단순하게 test라고 적는 것보다 describe - (context) - it 구조를 사용하여 테스트를 읽히도록 만들자

Commit:
삼항연산자를 이용하여 엘리먼트를 드러내라

- 삼항연산자로 엘리먼트를 드러내는 것보다 함수로 추출하는 게 좋다고 생각해서 이전에는 함수를 호출하는 형식으로 만들었음
- 하지만 함수로 엘리먼트를 감춰버리니 어떻게 렌더링되는지 한 눈에 알기가 어려워져서 삼항연산자를 이용해서 나타나도록 함

moonkii commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -4,57 +4,49 @@ import { useDispatch, useSelector } from 'react-redux';
+
+  beforeEach(() => {
+    useSelector.mockImplementation((selector) => selector({
+      ...given.restaurant,
```
`restaurant: given.restaurant` 같이 표현하시는게 코드 가독성에 더 좋을 것 같아요.

hanurii replied:
넵 다음 과제 때 반영해보겠니다!

moonkii commented:
과제를 잘 수행하셨어요!
회고 작성 후 공유해주시면 merge 하도록하겠습니다!

moonkii reviewed (approved):
과제를 잘 수행하셨어요!
회고 작성 후 공유해주시면 merge 하도록하겠습니다!
